target/linux/generic/files/drivers/platform/mikrotik/mikrotik_wlan_lz77.c

```diff
@@ -0,0 +1,443 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (C) 2023 John Thomson
+ */
+
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/string.h>
+#include <linux/errno.h>
+
+#include "mikrotik_wlan_lz77.h"
+
+#define MIKRO_LZ77 "[mikrotik wlan lz77] "
+
+/* examples have all decompressed to start with DRE\x00 */
+#define MIKRO_LZ77_EXPECTED_OUT	(('E' << 16) | ('R' << 8) | ('D'))
+
+/*
+ * The maximum number of bits used in a counter.
+ * For the look behind window, long instruction match offsets
+ * up to 6449 have been seen in provided compressed caldata blobs
+ * (that would need 21 counter bits: 4 to 12 + 11 to 0).
+ * conservative value here: 27 provides offset up to 0x8000 bytes
+ * uses a u8 in this code
+ */
+#define MIKRO_LZ77_MAX_COUNT_BIT_LEN 27
+
+enum mikrotik_wlan_lz77_instruction {
+	INSTR_ERROR = -1,
+	INSTR_LITERAL_BYTE = 0,
+	/* a (non aligned) byte follows this instruction,
+	 * which is directly copied into output
+	 */
+	INSTR_PREVIOUS_OFFSET = 1,
+	/* this group is a match, with a bytes length defined by
+	 * following counter bits, starting at bitshift 0,
+	 * less the built-in count of 1
+	 * using the previous offset as source
+	 */
+	INSTR_LONG = 2
+	/* this group has two counters,
+	 * the first counter starts at bitshift 4,
+	 *	 if this counter == 0, this is a non-matching group
+	 *	 the second counter (bytes length) starts at bitshift 4,
+	 *	 less the built-in count of 11+1.
+	 *	 The final match group has this count 0,
+	 *	 and following bits which pad to byte-alignment.
+	 *
+	 *	 if this counter > 0, this is a matching group
+	 *	 this first count is the match offset (in bytes)
+	 *	 the second count is the match length (in bytes),
+	 *	 less the built-in count of 2
+	 *	 these groups can source bytes that are part of this group
+	 */
+};
+
+struct lz77_mk_instr_opcodes {
+	/* group instruction */
+	enum mikrotik_wlan_lz77_instruction instruction;
+	/* if >0, a match group,
+	 * which starts at byte output_position - 1*offset
+	 */
+	size_t offset;
+	/* how long the match group is,
+	 * or how long the (following counter) non-match group is
+	 */
+	size_t length;
+	/* how many bits were used for this instruction + op code(s) */
+	size_t bits_used;
+	/* input char */
+	u8 *in;
+	/* offset where this instruction started */
+	size_t in_pos;
+};
+
+/**
+ * mikrotik_wlan_lz77_get_bit
+ *
+ * @in:			compressed data
+ * @in_offset_bit:	bit offset to extract
+ */
+static u8 mikrotik_wlan_lz77_get_bit(
+		const u8 *in,
+		size_t in_offset_bit)
+{
+	return ((in[in_offset_bit>>3] >> (in_offset_bit & 7)) & 0x1);
```

Comment:
the code description comment doesn't really explain the logic of what's going on here (why shift `in_offset_bit` by 3 to index the input and then shift the result by at most 3 bits). For someone not familiar with what's going on, this would probably need a little more explanation on what this function does and returns.

Reply:
commented, and used division and modulo directly, now:
```
 * convert the bit offset to byte offset,
 * shift to modulo of bits per bytes, so that wanted bit is lsb
 * and to extract only that bit.
 */
	return ((in[in_offset_bit / BITS_PER_BYTE] >>
		 (in_offset_bit % BITS_PER_BYTE)) &
		1);
```